commit 1c99e8d8e1d385b2d37311f9527f60c835010bdc
Author: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date:   Tue Jul 24 12:08:29 2018 +0100

    staging: bcm2835-camera: Ensure timestamps never go backwards.
    
    There is an awkward situation with H264 header bytes. Currently
    they are returned with a PTS of 0 because they aren't associated
    with a timestamped frame to encode. These are handled by either
    returning the timestamp of the last buffer to have been received,
    or in the case of the first buffer the timestamp taken at
    start_streaming.
    This results in a race where the current frame may have started
    before we take the start time, which results in the first encoded
    frame having an earlier timestamp than the header bytes.
    
    Ensure that we never return a negative delta to the user by checking
    against the previous timestamp.
    
    Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

diff --git a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
index 224291a1e997..f1fb11969a61 100644
--- a/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
+++ b/drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c
@@ -393,6 +393,11 @@ static void buffer_cb(struct vchiq_mmal_instance *instance,
 			 ktime_to_ns(dev->capture.kernel_start_ts),
 			 dev->capture.vc_start_timestamp, pts,
 			 ktime_to_ns(timestamp));
+		if (timestamp < dev->capture.last_timestamp) {
+			v4l2_dbg(1, bcm2835_v4l2_debug, &dev->v4l2_dev,
+				 "Negative delta - using last time\n");
+			timestamp = dev->capture.last_timestamp;
+		}
 		buf->vb.vb2_buf.timestamp = ktime_to_ns(timestamp);
 	} else {
 		if (dev->capture.last_timestamp) {
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1726_1c99e8d8e1d385b2d37311f9527f60c835010bdc.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1726_1c99e8d8e1d385b2d37311f9527f60c835010bdc.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
